gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+// mount is located by calling a mount lookup helper.
+func NewMountedFilesystemUpdater(rootDir string, ps *PositionedStructure, backupDir string, mountLookup mountLookupFunc) (*MountedFilesystemUpdater, error) {
+	fw, err := NewMountedFilesystemWriter(rootDir, ps)
+	if err != nil {
+		return nil, err
+	}
+	if mountLookup == nil {
+		return nil, fmt.Errorf("internal error: mount lookup helper must be provided")
+	}
+	if backupDir == "" {
+		return nil, fmt.Errorf("internal error: backup directory must not be unset")
+	}
+	fu := &MountedFilesystemUpdater{
+		MountedFilesystemWriter: fw,
+		backupDir:               backupDir,
+		mountLookup:             mountLookup,
+	}
+	return fu, nil
+}
+
+func fsStructBackupPath(backupDir string, ps *PositionedStructure) string {
+	return filepath.Join(backupDir, fmt.Sprintf("struct-%v", ps.Index))
+}
+
+// entryNames resolves the source, destination and backup locations for given
+// source/target combination. The source location is always within the root
+// directory passed during initialization. Backup location is within provided
+// backup directory or empty if directory was not provided.
+func (f *MountedFilesystemUpdater) entryNames(dstRoot, source, target, backupDir string) (srcName, dstName, backupName string) {
+	srcName = f.entrySource(source)
+
+	dstBase := target
+	if strings.HasSuffix(target, "/") {
+		// deploying to a directory
+		dstBase = filepath.Join(dstBase, filepath.Base(source))
+	}
+	dstName = filepath.Join(dstRoot, dstBase)
+
+	if backupDir != "" {
+		backupName = filepath.Join(backupDir, dstBase)
+	}
+
+	return srcName, dstName, backupName
+}
+
+// entrySource returns the location of given source within the root directory
+// provided during initialization.
+func (f *MountedFilesystemUpdater) entrySource(source string) string {
+	srcName := filepath.Join(f.contentDir, source)
+
+	if strings.HasSuffix(source, "/") {
+		// restore trailing / if one was there
+		srcName += "/"
+	}
+	return srcName
+}
+
+// Update applies an update to a mounted filesystem. The caller must have
+// executed a Backup() before, to prepare a data set for rollback purpose.
+func (f *MountedFilesystemUpdater) Update() error {
+	mount, err := f.mountLookup(f.ps)
+	if err != nil {
+		return fmt.Errorf("cannot find mount location of structure %v: %v", f.ps, err)
+	}
+
+	preserveInDst := prefixPreserve(mount, f.ps.Update.Preserve)
+	backupRoot := fsStructBackupPath(f.backupDir, f.ps)
+
+	for _, c := range f.ps.Content {
+		if err := f.updateVolumeContent(mount, &c, preserveInDst, backupRoot); err != nil {
+			return fmt.Errorf("cannot update content: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) updateDirectory(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName := f.entrySource(source)
+
+	if err := checkSrcIsDir(srcName); err != nil {
+		return fmt.Errorf("cannot update directory %q: %v", source, err)
+	}
+
+	fis, err := ioutil.ReadDir(srcName)
```

Comment:
we have a lot of repetition of this code up to 331 in the file

Reply:
Wrapped the check and listind the directory into a helper call.